Bloom.md

```diff
@@ -0,0 +1,248 @@
+* create operations (BF.ADD/MADD/INSERT): We can compute the estimated size of the bloom filter that will be created as a result of the operation. We can consider rejecting requests in case of unavailable memory.
+
+
+## Specification
+
+### Bloom Filter Command API
+
+The following Bloom Filter commands are supported and their API syntax is compatible with ReBloom:
+BF.ADD <key> <item>
+BF.EXISTS <key> <item>
+BF.MADD <key> <item> [<item> ...]
+BF.MEXISTS <key> <item> [<item> ...]
+BF.CARD <key>
+BF.INFO <key> [CAPACITY | SIZE | FILTERS | ITEMS | EXPANSION]
+BF.RESERVE <key> <false_positive_rate> <capacity> [EXPANSION <expansion>] | [NONSCALING]
+BF.INSERT <key> [CAPACITY <capacity>] [ERROR <fp_error>] [EXPANSION <expansion>] [NOCREATE] [NONSCALING] ITEMS <item> [<item> ...]
+
+Currently following commands (from ReBloom) are not supported:
+BF.LOADCHUNK <key> <iterator> <data>
+BF.SCANDUMP <key> <iterator>
+
+The BF.SCANDUMP command is used to perform an incremental save on specific Bloom filter object.
+The BF.LOADCHUNK is used to incrementally load / restore a bloom filter object from data from the BF.SCANDUMP command.
+
+The reason for not implementing these two commands is because the Module provides the ability to load and save BloomModule data type items during RDB load and save. BF.LOADCHUNK and BF.SCANDUMP are APIs to load bloom filter objects through commands, but since we will provide RDB save & load, having specific commands for the same purpose was not considered as required.
+
+Note: Based on the AOF Rewrite discussion, we will decide on whether these commands should be supported.
+
+### Configs
+
+The default properties using which Bloom Filter objects are created can be controlled using configs.
+
+The following module configurations are added to customize bloom filters:
+1. bf.bloom_capacity: Controls the default capacity. When create operations (BF.ADD/MADD) are used, bloom objects created will use the capacity specified by this config. This controls the number of items that can be added to a bloom filter object before it is scaled out (scaling) OR before it rejects add operations due to insufficient capacity on the object (nonscaling).
+2. bf.bloom_expansion_rate: Controls the default expansion rate. When create operations (BF.ADD/MADD) are used, bloom objects created will use the expansion rate specified by this config. This controls the capacity of the new filter that gets added to the list of filters as a result of scaling.
+3. bf.bloom_max_filters: Controls the maximum number of filters that any bloom object is allowed to scale out to.
+4. bf.bloom_fp_rate: Controls the default false positive rate that new bloom objects created (from BF.ADD/MADD) will use.
+5. bf.bloom_large_item_threshold: Memory usage of a bloom object beyond which bloom objects are exempted from defrag and are forced to be async freed (when deleted).
```

Comment:
```suggestion
1. bf.bloom-capacity: Controls the default capacity. When create operations (BF.ADD/MADD) are used, bloom objects created will use the capacity specified by this config. This controls the number of items that can be added to a bloom filter object before it is scaled out (scaling) OR before it rejects add operations due to insufficient capacity on the object (nonscaling).
2. bf.bloom-expansion-rate: Controls the default expansion rate. When create operations (BF.ADD/MADD) are used, bloom objects created will use the expansion rate specified by this config. This controls the capacity of the new filter that gets added to the list of filters as a result of scaling.
3. bf.bloom-max-filters: Controls the maximum number of filters that any bloom object is allowed to scale out to.
4. bf.bloom-fp-rate: Controls the default false positive rate that new bloom objects created (from BF.ADD/MADD) will use.
5. bf.bloom-large-item-threshold: Memory usage of a bloom object beyond which bloom objects are exempted from defrag and are forced to be async freed (when deleted).
```

Are these based of RedisBloom configs, or are these are own configs?
Why is 5 a configuration? That seems like something we should control.

Reply:
> Are these based of RedisBloom configs, or are these are own configs?

These are five configurations that are of own.

> Why is 5 a configuration? That seems like something we should control.

Do you mean we should have a static constant defined based on which we can handle large items differently (defrag, etc)? (Instead of making this configurable)

Reply:
> Do you mean we should have a static constant defined based on which we can handle large items differently (defrag, etc)? (Instead of making this configurable)

Yeah. I'm not familiar with other places where this is tunable.

Reply:
Sounds good. This can be changed to a static value. I will list some of the statics like this in a section below